action/container/ssh.go

```diff
@@ -0,0 +1,158 @@
+	TunnelReady chan (bool) // Channel to signal that SSH tunnel is not longer needed and can be closed
+}
+
+// Wait calls WaitFunc
+func (s OptsOpenSSH) Wait() {
+	s.WaitFunc()
+}
+
+// SettingsSSH is for functional option pattern
+type SettingsSSH func(*OptsOpenSSH)
+
+// WithWaitPressEnter is one possible function to pass into OpenSSH
+// It will wait until user presses ENTER key to shutdown the container
+func WithWaitPressEnter() SettingsSSH {
+	return func(s *OptsOpenSSH) {
+		s.WaitFunc = func() {
+			fmt.Print("Press ENTER to stop container ")
+			fmt.Scanln()
+			s.TunnelClose <- true
+		}
+	}
+}
+
+// WithWaitNone is one possible function to pass into OpenSSH
+// It will not wait
+func WithWaitNone() SettingsSSH {
+	return func(s *OptsOpenSSH) {
+		s.WaitFunc = func() {
+			fmt.Println("Skipping waiting")
+		}
+	}
+}
+
+// NewSettingsSSH returns a SettingsSSH
+func NewSettingsSSH(opts ...SettingsSSH) *OptsOpenSSH {
+	// Defaults
+	var m sync.Mutex
+	s := &OptsOpenSSH{
+		MutexData:   &m,
+		TunnelClose: make(chan (bool)),
+		TunnelReady: make(chan (bool)),
+	}
+	WithWaitPressEnter()(s)
+
+	for _, opt := range opts {
+		opt(s)
+	}
+	return s
+}
+
+// OpenSSH takes a container and starts SSH server with port exposed to the host
+func OpenSSH(
+	ctx context.Context,
+	client *dagger.Client,
+	container *dagger.Container,
+	workdir string,
+	opts *OptsOpenSSH,
+) error {
+	// Example in docs:
+	//   https://docs.dagger.io/cookbook/#expose-service-containers-to-host
+	// This feature is untested and instead relies on tears, blood and sweat produced during
+	//   it's development to work.
+	// UPDATE: After more tears, blood and sweat we also have some testing! Yippee!
+
+	if container == nil {
+		log.Println("skipping SSH because no container was given")
+		return nil
+	}
+
+	if workdir == "" {
+		workdir = "/"
+	}
+
+	// Generate a password for the root user
+	opts.MutexData.Lock()
+	opts.Password = generatePassword(16)
+	opts.MutexData.Unlock()
+
+	// Prepare the container
+	container = container.
+		WithExec([]string{"bash", "-c", fmt.Sprintf("echo 'root:%s' | chpasswd", opts.Password)}).
+		WithExec([]string{"bash", "-c", fmt.Sprintf("echo 'cd %s' >> /root/.bashrc", workdir)}).
+		WithExec([]string{"/usr/sbin/sshd", "-D"})
+
+	// Convert container to service with exposed SSH port
+	const sshPort = 22
+	sshServiceDoc := container.WithExposedPort(sshPort).AsService()
+
+	// Expose the SSH server to the host
+	sshServiceTunnel, err := client.Host().Tunnel(sshServiceDoc).Start(ctx)
+	if err != nil {
+		fmt.Println("Problem getting tunnel up")
+		return err
+	}
+	defer sshServiceTunnel.Stop(ctx) // nolint:errcheck
+
+	// Get and print instructions on how to connect
+	sshAddress, err := sshServiceTunnel.Endpoint(ctx)
+	if err != nil {
+		fmt.Println("problem getting address")
+		return err
+	}
+	sshAddressSplit := strings.Split(sshAddress, ":")
+	opts.Port = sshAddressSplit[1]
```

Comment:
Why are you setting the port here out side the lock and three lines further down inside the mutex lock?

Reply:
that is some copy-paste error since I set the port twice :D

Reply:
I will make this into function (method of the struct) and add some security to it.